bsip-0056.md

```diff
@@ -0,0 +1,283 @@
+balance/asset pair. 
+
+We here propose to add a new vesting policy:
+
+    struct managed_policy_initializer
+    {
+        fc::time_point_sec begin_timestamp;
+
+        // After begin_timestamp + vesting_cliff_seconds, the customer can withdraw the contained funds
+        // Must be >= 0
+        uint32_t           vesting_cliff_seconds = 0;
+    
+        // The delay in seconds it takes before a withdrawal transfers the funds to the owner account
+        // Must be >= 0
+        uint32_t              delay_seconds = 0;
+        
+        // An optional manager account that is allowed to perform settlements across vesting balances
+        optional<account_id_type>          manager;
+    }
+
+Note:
+ - Initiating a withdrawal means sending the withdraw operation to the blockchain
+ - A managed vesting balance can be created with `amount = 0`, a deposit operation will be introduced as well
+ - In order to avoid cluttering the in-ram state with unused vesting balances, the managed vesting balance will annul themselves after being unused for a specified time (see Annulment of vesting balances). Annullment means that the vesting balance object is removed and its funds are automatically transferred to the owner
+
+## Settlement Operation
+
+In this step, we allow the above `manager` to re-balance assets in corresponding vesting balances. The rebalancing
+takes the following form:
+
+    struct settlement
+    {
+        vesting_balance_id_type   vesting_balance;
+        share_type/asset          delta;
+    }
+
+When settling, a vector of `settlement` is used to redefine the amounts of the corresponding vesting balances.
+Obviously, the assets of the vesting balances need to match, and the sum of amounts *before* settlement needs to
+equal the sum of amounts *after* settlement. Generally, the settlement operation can move all available funds in the vesting balance, independent of how much the user can currently claim due to the vesting balance policy. 
+
+The settlement operation takes the following form:
+
+    struct vesting_balance_settle_operation : public base_operation
+    {
+       asset                       fee;
+       account_id_type             fee_paying_account;
+       vector<settlement>          settlements;
+
+       extension< ext > extensions;
+    };
+
+This operations overwrites the balances in the vesting balances according to the content of `settlements`.
+Obviously, a few conditions need to be met:
+
+1. The `manager` of each vesting balances must coincide with `fee_paying_account` of the `vesting_balance_settle_operation`
+2. There must be at least one `settlement` in `settlements` vector of `vesting_balance_settle_operation`
+3. A `settlement` can only contain non-zero `delta`s, must contain at least two `delta`s and all `delta`s in one `settlement` must target vesting balances with the same asset (`asset_id`)
```

Comment:
"at least two" makes the previous condition ("at least one") redundant.

Reply:
The settlement operation can settle several assets at the same time. Vector of settlements (at least one), each settlement contains several deltas (at least two, otherwise you can't have zero).

Reply:
Oh. That's not how I read it.
According to the definition above, a `settlement` contains a single delta, and a `vesting_balance_settle_operation` contains multiple deltas (`settlement`s) of only one asset type (*"the assets of the vesting balances need to match"*).

Reply:
Yes, because it was also stated incorrect in the BSIP and I didn't see it.

What I wrote above is how it should be, but it was not reflected in the pseudo-code. I fixed that now.

Reply:
This is the intended structure:

```
struct settlement
{
    vector<delta>   deltas;  // at least two, and the amounts within must sum up to zero
}

struct delta 
{
    vesting_balance_id_type  vesting_balance;  // vesting balance is uniquely assigned to an account
    share_type/asset         amount;           // delta amount can be positive or negative, but not zero
}

struct vesting_balance_settle_operation : public base_operation
{
   asset                       fee;
   account_id_type             fee_paying_account;
   vector<settlement>          settlements;

   extension< ext > extensions;
};
```

Would it make sense to add the `asset_id` to the `settlement`, and all settled `vesting_balance`s must contain same `asset_id`? This is merely for safety and clarity.

Reply:
Adding `asset_id` to `settlement` and using only `share_type` in `delta` would be more efficient, space-wise and logic-wise (no redundant asset_ids -> no need to check consistency).

Reply:
Done